core/src/avm2/object/xml_object.rs

```diff
@@ -177,6 +177,46 @@ impl<'gc> XmlObject<'gc> {
         Ref::map(self.0.read(), |data| &data.node)
     }
 
+    pub fn contains(&self, value: &Value<'gc>, activation: &mut Activation<'_, 'gc>) -> bool {
+        let node = self.node();
+
+        if let Some(xml) = value.as_object().and_then(|obj| obj.as_xml_object()) {
+            let other = xml.node();
+
+            if node.is_text() || node.is_attribute() {
+                return node.xml_to_string(activation) == other.xml_to_string(activation);
+            }
+
+            if other.is_text() || other.is_attribute() {
+                return false;
+            }
+
+            return node.equals(&other);
+        }
+
+        if let Some(list) = value.as_object().and_then(|obj| obj.as_xml_list_object()) {
+            if list.length() == 1 {
```

Comment:
I am not sure if FP actually considers single-element XMLLists here, further testing would be required. (Though looking at avmplus, they seem to just check for a XML object and use `_equals`)

Reply:
I did a test to see if FP supports single-element lists, and it looks like it does.
I hope I've answered your question correctly.

```as
<?xml version="1.0" encoding="utf-8"?>
<mx:Application xmlns:mx="http://www.adobe.com/2006/mxml" layout="absolute" minWidth="955" minHeight="600" creationComplete="init()">
	<mx:Script>
		<![CDATA[           
			private function init():void {
				// XML with single item
				var singleItemXML:XML = <order><item>Rice</item></order>;
				
				// Test if Flash Player considers single-element XML
				trace("Is singleItemXML a single element? " + singleItemXML.contains(<item>Rice</item>)); // true
				
				// Original XML with multiple items
				var xml:XML = 
					<order>
						<item>Rice</item>
						<item>Kung Pao Shrimp</item>
					</order>;
				trace("Is xml a single element? " + xml.contains(<item>Rice</item>)); // false
				
				// Test contains method with original XML
				trace(xml.item[0].contains(<item>Rice</item>)); // true
				trace(xml.item[1].contains(<item>Kung Pao Shrimp</item>)); // true
				trace(xml.item[1].contains(<item>MSG</item>)); // false
			}
		]]>
	</mx:Script>
</mx:Application>
```

FP result:
```
Is singleItemXML a single element? false
Is xml a single element? false
true
true
false
```

Ruffle result:
```
2024-05-04T09:36:59.023293Z  INFO avm_trace: Is singleItemXML a single element? false
2024-05-04T09:36:59.023318Z  INFO avm_trace: Is xml a single element? false
2024-05-04T09:36:59.023330Z  INFO avm_trace: true
2024-05-04T09:36:59.023336Z  INFO avm_trace: true
2024-05-04T09:36:59.023341Z  INFO avm_trace: false
```

Reply:
```actionscript
var xml:XML = <order><item>Rice</item><item>Kung Pao Shrimp</item></order>;

trace(xml.item[0].contains(<item>Rice</item>)); // true, both in Ruffle and FP

var list:XMLList = new XMLList("<item>Rice</item>");
trace(xml.item[0].contains(list)); // true in Ruffle with this PR, false in FP
```

So from my testing we do not need to handle XMLLists in the contains method at all, so we can safely remove this.

nit: I think it might be cleaner to move the contains implementation into the `globals/xml.rs` file, it really does not need to be on the XmlObject.

Reply:
I don't know how to move the `contains` implementation into the `globals/xml.rs` file, I let you do it, I'm not quite familiar with Rust syntax at the moment.

Anyway, according to our tests, the implementation of contains seems to work well!

Reply:
> I don't know how to move the `contains` implementation into the `globals/xml.rs` file, I let you do it, I'm not quite familiar with Rust syntax at the moment.

Something like this should work:
```rust
pub fn contains<'gc>(
    _activation: &mut Activation<'_, 'gc>,
    this: Object<'gc>,
    args: &[Value<'gc>],
) -> Result<Value<'gc>, Error<'gc>> {
    let xml = this.as_xml_object().unwrap();
    let value = args.get_value(0);

    if let Some(other) = value.as_object().and_then(|obj| obj.as_xml_object()) {
        let result = xml.node().equals(&other.node());
        return Ok(result.into());
    }
    
    Ok(false.into())
}
```

Reply:
That's right, it works, thank you! PR ready to be merged 😉